package net.minecraft.network.protocol.common;

import com.google.common.collect.Lists;
import java.util.List;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.PacketType;
import net.minecraft.network.protocol.common.custom.BrandPayload;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.network.protocol.common.custom.DiscardedPayload;
import net.minecraft.util.Util;

public record ClientboundCustomPayloadPacket(CustomPacketPayload payload) implements Packet<ClientCommonPacketListener> {
   private static final int MAX_PAYLOAD_SIZE = 1048576;
   public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundCustomPayloadPacket> GAMEPLAY_STREAM_CODEC;
   public static final StreamCodec<FriendlyByteBuf, ClientboundCustomPayloadPacket> CONFIG_STREAM_CODEC;

   public ClientboundCustomPayloadPacket(CustomPacketPayload var1) {
      super();
      this.payload = var1;
   }

   public PacketType<ClientboundCustomPayloadPacket> type() {
      return CommonPacketTypes.CLIENTBOUND_CUSTOM_PAYLOAD;
   }

   public void handle(ClientCommonPacketListener var1) {
      var1.handleCustomPayload(this);
   }

   static {
      GAMEPLAY_STREAM_CODEC = CustomPacketPayload.codec((CustomPacketPayload.FallbackProvider)((var0) -> DiscardedPayload.codec(var0, 1048576)), (List)Util.make(Lists.newArrayList(new CustomPacketPayload.TypeAndCodec[]{new CustomPacketPayload.TypeAndCodec(BrandPayload.TYPE, BrandPayload.STREAM_CODEC)}), (var0) -> {
      })).map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
      CONFIG_STREAM_CODEC = CustomPacketPayload.codec((CustomPacketPayload.FallbackProvider)((var0) -> DiscardedPayload.codec(var0, 1048576)), List.of(new CustomPacketPayload.TypeAndCodec(BrandPayload.TYPE, BrandPayload.STREAM_CODEC))).map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
   }
}
